In [10]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

In [11]:
event_data = pd.read_csv('MMADataset2023/ufc_event_data.csv')
display(event_data.head())
print("Shape of Event Data:", event_data.shape)
event_data.info()

event_id                                event_name  event_date event_city  \
0       665         UFC Fight Night: Dawson vs. Green  2023-10-07  Las Vegas   
1       664        UFC Fight Night: Fiziev vs. Gamrot  2023-09-23  Las Vegas   
2       663  UFC Fight Night: Grasso vs. Shevchenko 2  2023-09-16  Las Vegas   
3       662          UFC 293: Adesanya vs. Strickland  2023-09-09     Sydney   
4       661          UFC Fight Night: Gane vs. Spivac  2023-09-02      Paris   

       event_state event_country  \
0           Nevada           USA   
1           Nevada           USA   
2           Nevada           USA   
3  New South Wales     Australia   
4    Ile-de-France        France   

                                           event_url  
0  http://ufcstats.com/event-details/c8a49ff2acb6...  
1  http://ufcstats.com/event-details/c945adc22c2b...  
2  http://ufcstats.com/event-details/8fa2b0657236...  
3  http://ufcstats.com/event-details/ece280745f87...  
4  http://ufcstats.com/event-details/ef61d9f5176b...

Shape of Event Data: (665, 7)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 665 entries, 0 to 664
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   event_id       665 non-null    int64 
 1   event_name     665 non-null    object
 2   event_date     665 non-null    object
 3   event_city     665 non-null    object
 4   event_state    616 non-null    object
 5   event_country  665 non-null    object
 6   event_url      665 non-null    object
dtypes: int64(1), object(6)
memory usage: 36.5+ KB


In [12]:
fight_data = pd.read_csv('MMADataset2023/ufc_fight_data.csv')
display(fight_data.head())
print("Shape of Fight Data:", fight_data.shape)
fight_data.info()

fight_id  event_id          referee     f_1     f_2  winner num_rounds  \
0      7218       664        Herb Dean  2976.0  2884.0  2884.0          5   
1      7217       664       Mark Smith  1662.0  2464.0  1662.0          3   
2      7216       664     Kerry Hatley   981.0   179.0   981.0          3   
3      7215       664  Dan Miragliotta  3831.0  2974.0  3831.0          3   
4      7214       664        Herb Dean  1108.0  2320.0  2320.0          3   

  title_fight         weight_class gender      result  \
0           F          Lightweight      M      KO/TKO   
1           F        Featherweight      M    Decision   
2           F  Women's Strawweight      F      KO/TKO   
3           F         Welterweight      M  Submission   
4           F        Featherweight      M  Submission   

                       result_details  finish_round finish_time  \
0            to    \n      Leg Injury             2        2:03   
1                           Unanimous             3        5:00   
2          Punches to Head From Mount             2        2:42   
3                    Rear Naked Choke             2        4:32   
4  Guillotine Choke From Bottom Guard             1        3:12   

                                           fight_url  
0  http://ufcstats.com/fight-details/23a604f46028...  
1  http://ufcstats.com/fight-details/da1b37edb8cc...  
2  http://ufcstats.com/fight-details/d8335b728604...  
3  http://ufcstats.com/fight-details/bf647be41de3...  
4  http://ufcstats.com/fight-details/6e1bf1b163b3...

Shape of Fight Data: (7218, 15)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7218 entries, 0 to 7217
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   fight_id        7218 non-null   int64  
 1   event_id        7218 non-null   int64  
 2   referee         7186 non-null   object 
 3   f_1             7199 non-null   float64
 4   f_2             7205 non-null   float64
 5   winner          7203 non-null   float64
 6   num_rounds      7218 non-null   object 
 7   title_fight     7218 non-null   object 
 8   weight_class    7205 non-null   object 
 9   gender          7218 non-null   object 
 10  result          7218 non-null   object 
 11  result_details  7201 non-null   object 
 12  finish_round    7218 non-null   int64  
 13  finish_time     7218 non-null   object 
 14  fight_url       7218 non-null   object 
dtypes: float64(3), int64(3), object(9)
memory usage: 846.0+ KB


In [17]:
fight_stat_data = pd.read_csv('MMADataset2023/ufc_fight_stat_data.csv')
display(fight_stat_data.head())
print("Shape of Fight Stat Data:", fight_stat_data.shape)
fight_stat_data.info()


fight_stat_id  fight_id  fighter_id  knockdowns  total_strikes_att  \
0          14436      7218      2976.0         0.0               34.0   
1          14435      7218      2884.0         0.0               42.0   
2          14434      7217      1662.0         0.0               59.0   
3          14433      7217      2464.0         0.0               72.0   
4          14432      7216       981.0         0.0              130.0   

   total_strikes_succ  sig_strikes_att  sig_strikes_succ  takedown_att  \
0                19.0             32.0              18.0           0.0   
1                17.0             40.0              16.0           6.0   
2                37.0             40.0              23.0          15.0   
3                32.0             55.0              18.0           0.0   
4                90.0            102.0              70.0           1.0   

   takedown_succ  submission_att  reversals ctrl_time  \
0            0.0             0.0        0.0      0:00   
1            1.0             0.0        0.0      1:28   
2            5.0             1.0        0.0      7:33   
3            0.0             0.0        1.0      2:11   
4            1.0             0.0        0.0      2:03   

                                           fight_url  
0  http://ufcstats.com/fight-details/23a604f46028...  
1  http://ufcstats.com/fight-details/23a604f46028...  
2  http://ufcstats.com/fight-details/da1b37edb8cc...  
3  http://ufcstats.com/fight-details/da1b37edb8cc...  
4  http://ufcstats.com/fight-details/d8335b728604...

Shape of Fight Stat Data: (14436, 14)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14436 entries, 0 to 14435
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   fight_stat_id       14436 non-null  int64  
 1   fight_id            14436 non-null  int64  
 2   fighter_id          14404 non-null  float64
 3   knockdowns          14394 non-null  float64
 4   total_strikes_att   14394 non-null  float64
 5   total_strikes_succ  14394 non-null  float64
 6   sig_strikes_att     14394 non-null  float64
 7   sig_strikes_succ    14394 non-null  float64
 8   takedown_att        14394 non-null  float64
 9   takedown_succ       14394 non-null  float64
 10  submission_att      14394 non-null  float64
 11  reversals           14394 non-null  float64
 12  ctrl_time           14394 non-null  object 
 13  fight_url           14436 non-null  object 
dtypes: float64(10), int64(2), object(2)
memory usage: 1.5+ MB


In [18]:
fighter_data = pd.read_csv('MMADataset2023/ufc_fighter_data.csv')
display(fighter_data.head())
print("Shape of Fighter Data:", fighter_data.shape)
fighter_data.info()


fighter_id fighter_f_name fighter_l_name fighter_nickname  \
0        4107            Tom          Aaron              NaN   
1        4106          Danny         Abbadi     The Assassin   
2        4105        Nariman        Abbasov        Bayraktar   
3        4104          David         Abbott             Tank   
4        4103          Hamdy     Abdelwahab       The Hammer   

   fighter_height_cm  fighter_weight_lbs  fighter_reach_cm fighter_stance  \
0                NaN               155.0               NaN            NaN   
1             180.34               155.0               NaN       Orthodox   
2             172.72               155.0            167.64       Orthodox   
3             182.88               265.0               NaN         Switch   
4             187.96               264.0            182.88       Southpaw   

  fighter_dob  fighter_w  fighter_l  fighter_d  fighter_nc_dq  \
0  1978-07-13          5          3          0            NaN   
1  1983-07-03          4          6          0            NaN   
2  1994-02-01         28          4          0            NaN   
3         NaN         10         15          0            NaN   
4  1993-01-22          5          0          0            1.0   

                                         fighter_url  
0  http://ufcstats.com/fighter-details/93fe7332d1...  
1  http://ufcstats.com/fighter-details/15df64c02b...  
2  http://ufcstats.com/fighter-details/59a9d6dac6...  
3  http://ufcstats.com/fighter-details/b361180739...  
4  http://ufcstats.com/fighter-details/3329d692ae...

Shape of Fighter Data: (4107, 14)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4107 entries, 0 to 4106
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   fighter_id          4107 non-null   int64  
 1   fighter_f_name      4107 non-null   object 
 2   fighter_l_name      4092 non-null   object 
 3   fighter_nickname    2250 non-null   object 
 4   fighter_height_cm   3797 non-null   float64
 5   fighter_weight_lbs  4020 non-null   float64
 6   fighter_reach_cm    2166 non-null   float64
 7   fighter_stance      3273 non-null   object 
 8   fighter_dob         3349 non-null   object 
 9   fighter_w           4107 non-null   int64  
 10  fighter_l           4107 non-null   int64  
 11  fighter_d           4107 non-null   int64  
 12  fighter_nc_dq       482 non-null    float64
 13  fighter_url         4107 non-null   object 
dtypes: float64(4), int64(4), object(6)
memory usage: 449.3+ KB


In [19]:
master_dataframe = pd.read_csv('UFCMMABiggestDatasetWithDifferentials/masterdataframe.csv')
display(master_dataframe.head())
print("Shape of Master Dataframe:", master_dataframe.shape)
master_dataframe.info()

date                                          fight_url  \
0  1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
1  1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
2  1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
3  1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
4  1994-03-11  http://ufcstats.com/fight-details/ccee020be2e8...   

                                           event_url  result         fighter  \
0  http://ufcstats.com/event-details/a6a9ab5a824e...       0  Sean Daugherty   
1  http://ufcstats.com/event-details/a6a9ab5a824e...       1    Scott Morris   
2  http://ufcstats.com/event-details/a6a9ab5a824e...       0      Ray Wizard   
3  http://ufcstats.com/event-details/a6a9ab5a824e...       1   Patrick Smith   
4  http://ufcstats.com/event-details/a6a9ab5a824e...       0   David Levicki   

         opponent     division    stance         dob  method  ...  \
0    Scott Morris  Open Weight       NaN  1975-12-04     SUB  ...   
1  Sean Daugherty  Open Weight  Orthodox         NaN     SUB  ...   
2   Patrick Smith  Open Weight       NaN         NaN     SUB  ...   
3      Ray Wizard  Open Weight  Orthodox  1963-08-28     SUB  ...   
4   Johnny Rhodes  Open Weight       NaN         NaN  KO/TKO  ...   

   recent_avg_clinch_strikes_attempts_per_min  \
0                                         NaN   
1                                         NaN   
2                                         NaN   
3                                         NaN   
4                                         NaN   

   precomp_recent_avg_clinch_strikes_attempts_per_min  \
0                                                NaN    
1                                                NaN    
2                                                NaN    
3                                                NaN    
4                                                NaN    

  avg_ground_strikes_landed_per_min precomp_avg_ground_strikes_landed_per_min  \
0                          0.000000                                       NaN   
1                          0.000000                                       NaN   
2                          0.000000                                       NaN   
3                          0.000000                                       NaN   
4                          0.081855                                       NaN   

  recent_avg_ground_strikes_landed_per_min  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   

  precomp_recent_avg_ground_strikes_landed_per_min  \
0                                              NaN   
1                                              NaN   
2                                              NaN   
3                                              NaN   
4                                              NaN   

  avg_ground_strikes_attempts_per_min  \
0                            0.000000   
1                            0.000000   
2                            0.000000   
3                            0.000000   
4                            0.081855   

   precomp_avg_ground_strikes_attempts_per_min  \
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   

   recent_avg_ground_strikes_attempts_per_min  \
0                                         NaN   
1                                         NaN   
2                                         NaN   
3                                         NaN   
4                                         NaN   

   precomp_recent_avg_ground_strikes_attempts_per_min  
0                                                NaN   
1                     

Shape of Master Dataframe: (13322, 530)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13322 entries, 0 to 13321
Columns: 530 entries, date to precomp_recent_avg_ground_strikes_attempts_per_min
dtypes: float64(491), int64(25), object(14)
memory usage: 53.9+ MB


In [20]:
print('Event Data shape:', event_data.shape)
print('Fight Data shape:', fight_data.shape)
print('Fight Stat Data shape:', fight_stat_data.shape)
print('Fighter Data shape:', fighter_data.shape)
print('Master Dataframe shape:', master_dataframe.shape)

Event Data shape: (665, 7)
Fight Data shape: (7218, 15)
Fight Stat Data shape: (14436, 14)
Fighter Data shape: (4107, 14)
Master Dataframe shape: (13322, 530)


In [22]:
print("\nEvent Data Summary:")
display(event_data.describe())
print("\nFight Data Summary:")
display(fight_data.describe())
print("\nFight Stat Data Summary:")
display(fight_stat_data.describe())
print("\nFighter Data Summary:")
display(fighter_data.describe())
print("\nMaster Dataframe Summary:")
display(master_dataframe.describe())


Event Data Summary:


event_id
count  665.000000
mean   333.000000
std    192.113248
min      1.000000
25%    167.000000
50%    333.000000
75%    499.000000
max    665.000000


Fight Data Summary:


fight_id     event_id          f_1          f_2       winner  \
count  7218.000000  7218.000000  7199.000000  7205.000000  7203.000000   
mean   3609.500000   349.977418  2072.702042  2072.142124  2074.013467   
std    2083.801454   186.567754  1174.571035  1171.522450  1168.685492   
min       1.000000     1.000000     7.000000     2.000000     7.000000   
25%    1805.250000   193.250000  1070.000000  1065.000000  1092.000000   
50%    3609.500000   355.000000  2034.000000  2070.000000  2050.000000   
75%    5413.750000   508.750000  3102.000000  3102.000000  3103.000000   
max    7218.000000   664.000000  4104.000000  4106.000000  4104.000000   

       finish_round  
count   7218.000000  
mean       2.332086  
std        1.014717  
min        1.000000  
25%        1.000000  
50%        3.000000  
75%        3.000000  
max        5.000000


Fight Stat Data Summary:


fight_stat_id      fight_id    fighter_id    knockdowns  \
count   14436.000000  14436.000000  14404.000000  14394.000000   
mean     7218.500000   3609.500000   2072.421966      0.215506   
std      4167.458578   2083.729274   1173.006408      0.493373   
min         1.000000      1.000000      2.000000      0.000000   
25%      3609.750000   1805.000000   1066.000000      0.000000   
50%      7218.500000   3609.500000   2052.000000      0.000000   
75%     10827.250000   5414.000000   3102.000000      0.000000   
max     14436.000000   7218.000000   4106.000000      5.000000   

       total_strikes_att  total_strikes_succ  sig_strikes_att  \
count       14394.000000        14394.000000     14394.000000   
mean          101.137002           53.620050        80.818883   
std            77.890569           43.802148        69.920549   
min             0.000000            0.000000         0.000000   
25%            36.000000           19.000000        26.000000   
50%            88.000000           45.000000        63.000000   
75%           149.000000           78.000000       117.000000   
max           746.000000          447.000000       744.000000   

       sig_strikes_succ  takedown_att  takedown_succ  submission_att  \
count      14394.000000  14394.000000   14394.000000    14394.000000   
mean          35.746283      2.800195       1.058774        0.393428   
std           31.738411      3.712215       1.683542        0.839972   
min            0.000000      0.000000       0.000000        0.000000   
25%           12.000000      0.000000       0.000000        0.000000   
50%           28.000000      1.000000       0.000000        0.000000   
75%           51.000000      4.000000       1.000000        0.000000   
max          445.000000     49.000000      21.000000       10.000000   

          reversals  
count  14394.000000  
mean       0.134570  
std        0.422097  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        6.000000


Fighter Data Summary:


fighter_id  fighter_height_cm  fighter_weight_lbs  fighter_reach_cm  \
count  4107.000000        3797.000000         4020.000000       2166.000000   
mean   2054.000000         178.257561          170.645522        181.841016   
std    1185.733107           8.887685           39.654222         10.684047   
min       1.000000         152.400000          105.000000        147.320000   
25%    1027.500000         172.720000          145.000000        175.260000   
50%    2054.000000         177.800000          170.000000        182.880000   
75%    3080.500000         185.420000          185.000000        190.500000   
max    4107.000000         226.060000          770.000000        213.360000   

         fighter_w    fighter_l    fighter_d  fighter_nc_dq  
count  4107.000000  4107.000000  4107.000000     482.000000  
mean     12.346725     5.707816     0.252009       1.134855  
std       9.374810     5.110169     0.742583       0.387525  
min       0.000000     0.000000     0.000000       1.000000  
25%       6.000000     2.000000     0.000000       1.000000  
50%      11.000000     4.000000     0.000000       1.000000  
75%      17.000000     8.000000     0.000000       1.000000  
max     253.000000    83.000000     9.000000       4.000000


Master Dataframe Summary:


result  total_comp_time         round         reach  \
count  13322.000000     13322.000000  13322.000000  12039.000000   
mean       0.491142       628.426813      2.330431     72.136307   
std        0.499940       365.103370      1.013957      4.176705   
min        0.000000         5.000000      1.000000     58.000000   
25%        0.000000       272.000000      1.000000     70.000000   
50%        0.000000       794.000000      3.000000     72.000000   
75%        1.000000       900.000000      3.000000     75.000000   
max        1.000000      2160.000000      5.000000     84.000000   

             height           age    knockdowns  sub_attempts     reversals  \
count  13308.000000  13087.000000  13322.000000  13322.000000  13322.000000   
mean      70.383228     37.109345      0.215283      0.402792      0.136991   
std        3.482012      6.217236      0.493830      0.852478      0.423131   
min       60.000000     21.000000      0.000000      0.000000      0.000000   
25%       68.000000     33.000000      0.000000      0.000000      0.000000   
50%       71.000000     37.000000      0.000000      0.000000      0.000000   
75%       73.000000     41.000000      0.000000      1.000000      0.000000   
max       83.000000     79.000000      5.000000     10.000000      5.000000   

            control  ...  recent_avg_clinch_strikes_attempts_per_min  \
count  13322.000000  ...                                 9179.000000   
mean     131.914953  ...                                    0.809267   
std      174.784787  ...                                    0.954325   
min        0.000000  ...                                    0.000000   
25%        5.000000  ...                                    0.284096   
50%       57.000000  ...                                    0.568116   
75%      195.000000  ...                                    1.017945   
max     1342.000000  ...                                   16.196825   

       precomp_recent_avg_clinch_strikes_attempts_per_min  \
count                                        7689.000000    
mean                                            0.833145    
std                                             0.972702    
min                                             0.000000    
25%                                             0.288889    
50%                                             0.587569    
75%                                             1.045211    
max                                            16.196825    

       avg_ground_strikes_landed_per_min  \
count                       13322.000000   
mean                            0.909553   
std                             1.587399   
min                             0.000000   
25%                             0.200000   
50%                             0.535997   
75%                             1.078072   
max                            48.000000   

       precomp_avg_ground_strikes_landed_per_min  \
count                               11061.000000   
mean                                    0.976845   
std                                     1.687217   
min                                     0.000000   
25%                                     0.233333   
50%                                     0.592045   
75%                                     1.153730   
max                                    48.000000   

       recent_avg_ground_strikes_landed_per_min  \
count                               9179.000000   
mean                                   0.802686   
std                                    1.220934   
min                                    0.000000   
25%                                    0.142540   
50%                                    0.422222   
75%                                    0.978280   
max                                   16.170940   

       precomp_recent_avg_ground_strikes_landed_per_min  \
count                                       7689.000000   
mean                                         

In [23]:
# Preview the first few rows of each dataset
print("UFC Event Data:")
display(event_data.head())

print("\nUFC Fight Data:")
display(fight_data.head())

print("\nUFC Fight Stat Data:")
display(fight_stat_data.head())

print("\nUFC Fighter Data:")
display(fighter_data.head())

print("\nMaster Dataframe:")
display(master_dataframe.head())


UFC Event Data:


event_id                                event_name  event_date event_city  \
0       665         UFC Fight Night: Dawson vs. Green  2023-10-07  Las Vegas   
1       664        UFC Fight Night: Fiziev vs. Gamrot  2023-09-23  Las Vegas   
2       663  UFC Fight Night: Grasso vs. Shevchenko 2  2023-09-16  Las Vegas   
3       662          UFC 293: Adesanya vs. Strickland  2023-09-09     Sydney   
4       661          UFC Fight Night: Gane vs. Spivac  2023-09-02      Paris   

       event_state event_country  \
0           Nevada           USA   
1           Nevada           USA   
2           Nevada           USA   
3  New South Wales     Australia   
4    Ile-de-France        France   

                                           event_url  
0  http://ufcstats.com/event-details/c8a49ff2acb6...  
1  http://ufcstats.com/event-details/c945adc22c2b...  
2  http://ufcstats.com/event-details/8fa2b0657236...  
3  http://ufcstats.com/event-details/ece280745f87...  
4  http://ufcstats.com/event-details/ef61d9f5176b...


UFC Fight Data:


fight_id  event_id          referee     f_1     f_2  winner num_rounds  \
0      7218       664        Herb Dean  2976.0  2884.0  2884.0          5   
1      7217       664       Mark Smith  1662.0  2464.0  1662.0          3   
2      7216       664     Kerry Hatley   981.0   179.0   981.0          3   
3      7215       664  Dan Miragliotta  3831.0  2974.0  3831.0          3   
4      7214       664        Herb Dean  1108.0  2320.0  2320.0          3   

  title_fight         weight_class gender      result  \
0           F          Lightweight      M      KO/TKO   
1           F        Featherweight      M    Decision   
2           F  Women's Strawweight      F      KO/TKO   
3           F         Welterweight      M  Submission   
4           F        Featherweight      M  Submission   

                       result_details  finish_round finish_time  \
0            to    \n      Leg Injury             2        2:03   
1                           Unanimous             3        5:00   
2          Punches to Head From Mount             2        2:42   
3                    Rear Naked Choke             2        4:32   
4  Guillotine Choke From Bottom Guard             1        3:12   

                                           fight_url  
0  http://ufcstats.com/fight-details/23a604f46028...  
1  http://ufcstats.com/fight-details/da1b37edb8cc...  
2  http://ufcstats.com/fight-details/d8335b728604...  
3  http://ufcstats.com/fight-details/bf647be41de3...  
4  http://ufcstats.com/fight-details/6e1bf1b163b3...


UFC Fight Stat Data:


fight_stat_id  fight_id  fighter_id  knockdowns  total_strikes_att  \
0          14436      7218      2976.0         0.0               34.0   
1          14435      7218      2884.0         0.0               42.0   
2          14434      7217      1662.0         0.0               59.0   
3          14433      7217      2464.0         0.0               72.0   
4          14432      7216       981.0         0.0              130.0   

   total_strikes_succ  sig_strikes_att  sig_strikes_succ  takedown_att  \
0                19.0             32.0              18.0           0.0   
1                17.0             40.0              16.0           6.0   
2                37.0             40.0              23.0          15.0   
3                32.0             55.0              18.0           0.0   
4                90.0            102.0              70.0           1.0   

   takedown_succ  submission_att  reversals ctrl_time  \
0            0.0             0.0        0.0      0:00   
1            1.0             0.0        0.0      1:28   
2            5.0             1.0        0.0      7:33   
3            0.0             0.0        1.0      2:11   
4            1.0             0.0        0.0      2:03   

                                           fight_url  
0  http://ufcstats.com/fight-details/23a604f46028...  
1  http://ufcstats.com/fight-details/23a604f46028...  
2  http://ufcstats.com/fight-details/da1b37edb8cc...  
3  http://ufcstats.com/fight-details/da1b37edb8cc...  
4  http://ufcstats.com/fight-details/d8335b728604...


UFC Fighter Data:


fighter_id fighter_f_name fighter_l_name fighter_nickname  \
0        4107            Tom          Aaron              NaN   
1        4106          Danny         Abbadi     The Assassin   
2        4105        Nariman        Abbasov        Bayraktar   
3        4104          David         Abbott             Tank   
4        4103          Hamdy     Abdelwahab       The Hammer   

   fighter_height_cm  fighter_weight_lbs  fighter_reach_cm fighter_stance  \
0                NaN               155.0               NaN            NaN   
1             180.34               155.0               NaN       Orthodox   
2             172.72               155.0            167.64       Orthodox   
3             182.88               265.0               NaN         Switch   
4             187.96               264.0            182.88       Southpaw   

  fighter_dob  fighter_w  fighter_l  fighter_d  fighter_nc_dq  \
0  1978-07-13          5          3          0            NaN   
1  1983-07-03          4          6          0            NaN   
2  1994-02-01         28          4          0            NaN   
3         NaN         10         15          0            NaN   
4  1993-01-22          5          0          0            1.0   

                                         fighter_url  
0  http://ufcstats.com/fighter-details/93fe7332d1...  
1  http://ufcstats.com/fighter-details/15df64c02b...  
2  http://ufcstats.com/fighter-details/59a9d6dac6...  
3  http://ufcstats.com/fighter-details/b361180739...  
4  http://ufcstats.com/fighter-details/3329d692ae...


Master Dataframe:


date                                          fight_url  \
0  1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
1  1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
2  1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
3  1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
4  1994-03-11  http://ufcstats.com/fight-details/ccee020be2e8...   

                                           event_url  result         fighter  \
0  http://ufcstats.com/event-details/a6a9ab5a824e...       0  Sean Daugherty   
1  http://ufcstats.com/event-details/a6a9ab5a824e...       1    Scott Morris   
2  http://ufcstats.com/event-details/a6a9ab5a824e...       0      Ray Wizard   
3  http://ufcstats.com/event-details/a6a9ab5a824e...       1   Patrick Smith   
4  http://ufcstats.com/event-details/a6a9ab5a824e...       0   David Levicki   

         opponent     division    stance         dob  method  ...  \
0    Scott Morris  Open Weight       NaN  1975-12-04     SUB  ...   
1  Sean Daugherty  Open Weight  Orthodox         NaN     SUB  ...   
2   Patrick Smith  Open Weight       NaN         NaN     SUB  ...   
3      Ray Wizard  Open Weight  Orthodox  1963-08-28     SUB  ...   
4   Johnny Rhodes  Open Weight       NaN         NaN  KO/TKO  ...   

   recent_avg_clinch_strikes_attempts_per_min  \
0                                         NaN   
1                                         NaN   
2                                         NaN   
3                                         NaN   
4                                         NaN   

   precomp_recent_avg_clinch_strikes_attempts_per_min  \
0                                                NaN    
1                                                NaN    
2                                                NaN    
3                                                NaN    
4                                                NaN    

  avg_ground_strikes_landed_per_min precomp_avg_ground_strikes_landed_per_min  \
0                          0.000000                                       NaN   
1                          0.000000                                       NaN   
2                          0.000000                                       NaN   
3                          0.000000                                       NaN   
4                          0.081855                                       NaN   

  recent_avg_ground_strikes_landed_per_min  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   

  precomp_recent_avg_ground_strikes_landed_per_min  \
0                                              NaN   
1                                              NaN   
2                                              NaN   
3                                              NaN   
4                                              NaN   

  avg_ground_strikes_attempts_per_min  \
0                            0.000000   
1                            0.000000   
2                            0.000000   
3                            0.000000   
4                            0.081855   

   precomp_avg_ground_strikes_attempts_per_min  \
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   

   recent_avg_ground_strikes_attempts_per_min  \
0                                         NaN   
1                                         NaN   
2                                         NaN   
3                                         NaN   
4                                         NaN   

   precomp_recent_avg_ground_strikes_attempts_per_min  
0                                                NaN   
1                     

Dataset Shapes: 
    Event Data shape: (665, 7)
    Fight Data shape: (7218, 15)
    Fight Stat Data shape: (14436, 14)
    Fighter Data shape: (4107, 14)
    Master Dataframe shape: (13322, 530)
Insights:
    UFC Event Data: 
        Columns: 7 (all non-null except event_state)
        Data Types: Mostly object(string) except event_id(int)
        Missing Values: Some missing event_state
        
    UFC Fight Data:
        Columns: 15 (floats, int, object)
        Missing Values: Some missing referee and weight_class 
        
    UFC Fight Stat Data:
        Columns: 14, mostly nums
        Missing Values: fighter_id and knockdowns have some missing values
        
    UFC Fighter Data:
        Columns: 14, nums but some categorical like stance
        Missing Values: Minimal
        
    Master Dataframe:
        Columns: 530 
        I think some columns are sparse or repetitive. 

### Missing Values

In [35]:
print("\nMissing Values in Event Data:")
print(event_data.isnull().sum())

print("\nMissing Values in Fight Data:")
print(fight_data.isnull().sum())

print("\nMissing Values in Fight Stat Data:")
print(fight_stat_data.isnull().sum())

print("\nMissing Values in Fighter Data:")
print(fighter_data.isnull().sum())



Missing Values in Event Data:
event_id         0
event_name       0
event_date       0
event_city       0
event_state      0
event_country    0
event_url        0
dtype: int64

Missing Values in Fight Data:
fight_id          0
event_id          0
referee           0
f_1               0
f_2               0
winner            0
num_rounds        0
title_fight       0
weight_class      0
gender            0
result            0
result_details    0
finish_round      0
finish_time       0
fight_url         0
dtype: int64

Missing Values in Fight Stat Data:
fight_stat_id         0
fight_id              0
fighter_id            0
knockdowns            0
total_strikes_att     0
total_strikes_succ    0
sig_strikes_att       0
sig_strikes_succ      0
takedown_att          0
takedown_succ         0
submission_att        0
reversals             0
ctrl_time             0
fight_url             0
dtype: int64

Missing Values in Fighter Data:
fighter_id            0
fighter_f_name        0
fighter_l_nam

event_data: "Unknown" since this isn't critical information in my opinion

In [36]:
event_data['event_state'].fillna('Unknown', inplace=True)

referee, weight_class, result_details: "Not Available"

f_1, f_2, winner: Fill with 0 or drop rows? Probably drop

In [37]:
fight_data[['referee', 'weight_class', 'result_details']] = fight_data[
    ['referee', 'weight_class', 'result_details']].fillna('Not Available')
fight_data[['f_1', 'f_2', 'winner']] = fight_data[['f_1', 'f_2', 'winner']].fillna(0)

fighter_id: Fill with 0 because we probably do not need this, but we can drop if needed. 
fight stats (knockdowns, total_strikes_att): Fill with 0

In [38]:
fight_stat_data['fighter_id'].fillna(0, inplace=True)
fight_stat_data.fillna(0, inplace=True)

fighter_l_name: Unknown

fighter_nickname, fighter_stance: Not Available

fighter_height_cm, fighter_weight_lbs, fighter_reach_cm, fighter_dob: maybe median value? probably drop 

fighter_nc_dq: Fill with 0

In [39]:
fighter_data['fighter_l_name'].fillna('Unknown', inplace=True)
fighter_data[['fighter_nickname', 'fighter_stance']] = fighter_data[
    ['fighter_nickname', 'fighter_stance']].fillna('Not Available')

fighter_data['fighter_height_cm'].fillna(fighter_data['fighter_height_cm'].median(), inplace=True)
fighter_data['fighter_weight_lbs'].fillna(fighter_data['fighter_weight_lbs'].median(), inplace=True)
fighter_data['fighter_reach_cm'].fillna(fighter_data['fighter_reach_cm'].median(), inplace=True)

fighter_data['fighter_dob'].fillna('1900-01-01', inplace=True)

fighter_data['fighter_nc_dq'].fillna(0, inplace=True)

Master Dataframe has about 318 columns of missing values so have not checked this. 

### Verifying missing values are handled

In [41]:
print("\nMissing Values in Event Data:")
print(event_data.isnull().sum())

print("\nMissing Values in Fight Data:")
print(fight_data.isnull().sum())

print("\nMissing Values in Fight Stat Data:")
print(fight_stat_data.isnull().sum())

print("\nMissing Values in Fighter Data:")
print(fighter_data.isnull().sum())



Missing Values in Event Data:
event_id         0
event_name       0
event_date       0
event_city       0
event_state      0
event_country    0
event_url        0
dtype: int64

Missing Values in Fight Data:
fight_id          0
event_id          0
referee           0
f_1               0
f_2               0
winner            0
num_rounds        0
title_fight       0
weight_class      0
gender            0
result            0
result_details    0
finish_round      0
finish_time       0
fight_url         0
dtype: int64

Missing Values in Fight Stat Data:
fight_stat_id         0
fight_id              0
fighter_id            0
knockdowns            0
total_strikes_att     0
total_strikes_succ    0
sig_strikes_att       0
sig_strikes_succ      0
takedown_att          0
takedown_succ         0
submission_att        0
reversals             0
ctrl_time             0
fight_url             0
dtype: int64

Missing Values in Fighter Data:
fighter_id            0
fighter_f_name        0
fighter_l_nam

### Checking for Duplicates

In [43]:
print('Duplicate rows in Event Data:', event_data.duplicated().sum())
print('Duplicate rows in Fight Data:', fight_data.duplicated().sum())
print('Duplicate rows in Fight Stat Data:', fight_stat_data.duplicated().sum())
print('Duplicate rows in Fighter Data:', fighter_data.duplicated().sum())
print('Duplicate rows in Master Dataframe:', master_dataframe.duplicated().sum())

Duplicate rows in Event Data: 0
Duplicate rows in Fight Data: 0
Duplicate rows in Fight Stat Data: 0
Duplicate rows in Fighter Data: 0
Duplicate rows in Master Dataframe: 0


In [45]:
print('Duplicate fight_id in Fight Data:', fight_data.duplicated(subset=['fight_id']).sum())
print('Duplicate fight_id in Fight Stat Data:', fight_stat_data.duplicated(subset=['fight_id']).sum())
print('Duplicate fighter_id in Fighter Data:', fighter_data.duplicated(subset=['fighter_id']).sum())
print('Duplicate rows in Master Dataframe:', master_dataframe.duplicated().sum())

Duplicate fight_id in Fight Data: 0
Duplicate fight_id in Fight Stat Data: 7218
Duplicate fighter_id in Fighter Data: 0
Duplicate rows in Master Dataframe: 0


In case there are duplicates or partial duplicates (keep first occurrence)

In [46]:
event_data = event_data.drop_duplicates()
fight_data = fight_data.drop_duplicates()
fight_stat_data = fight_stat_data.drop_duplicates()
fighter_data = fighter_data.drop_duplicates()
master_dataframe = master_dataframe.drop_duplicates()

In [48]:
fight_data = fight_data.drop_duplicates(subset=['fight_id'], keep='first')
fighter_data = fighter_data.drop_duplicates(subset=['fighter_id'], keep='first')

### Standardize Column Names

In [53]:
event_data.columns = event_data.columns.str.lower().str.replace(' ', '_')
fight_data.columns = fight_data.columns.str.lower().str.replace(' ', '_')
fight_stat_data.columns = fight_stat_data.columns.str.lower().str.replace(' ', '_')
fighter_data.columns = fighter_data.columns.str.lower().str.replace(' ', '_')
master_dataframe.columns = master_dataframe.columns.str.lower().str.replace(' ', '_')

In [54]:
event_data_columns = event_data.columns
fight_data_columns = fight_data.columns
fight_stat_data_columns = fight_stat_data.columns
fighter_data_columns = fighter_data.columns
master_dataframe_columns = master_dataframe.columns

event_data_columns, fight_data_columns, fight_stat_data_columns, fighter_data_columns, master_dataframe_columns

(Index(['event_id', 'event_name', 'event_date', 'event_city', 'event_state',
        'event_country', 'event_url'],
       dtype='object'),
 Index(['fight_id', 'event_id', 'referee', 'f_1', 'f_2', 'winner', 'num_rounds',
        'title_fight', 'weight_class', 'gender', 'result', 'result_details',
        'finish_round', 'finish_time', 'fight_url'],
       dtype='object'),
 Index(['fight_stat_id', 'fight_id', 'fighter_id', 'knockdowns',
        'total_strikes_att', 'total_strikes_succ', 'sig_strikes_att',
        'sig_strikes_succ', 'takedown_att', 'takedown_succ', 'submission_att',
        'reversals', 'ctrl_time', 'fight_url'],
       dtype='object'),
 Index(['fighter_id', 'fighter_f_name', 'fighter_l_name', 'fighter_nickname',
        'fighter_height_cm', 'fighter_weight_lbs', 'fighter_reach_cm',
        'fighter_stance', 'fighter_dob', 'fighter_w', 'fighter_l', 'fighter_d',
        'fighter_nc_dq', 'fighter_url'],
       dtype='object'),
 Index(['date', 'fight_url', 'event_url', 're

### Converting Data Type
event_date in Event Data = datetime64[ns]

fighter_dob in Fighter Data = datetime64[ns]

date and dob in Master Dataframe = datetime64[ns]

f_1, f_2, winner in Fight Data = float64

knockdowns, total_strikes_att, submission_att in Fight Stat Data = float64

In [56]:
event_data['event_date'] = pd.to_datetime(event_data['event_date'], errors='coerce')
fighter_data['fighter_dob'] = pd.to_datetime(fighter_data['fighter_dob'], errors='coerce')
master_dataframe['date'] = pd.to_datetime(master_dataframe['date'], errors='coerce')
master_dataframe['dob'] = pd.to_datetime(master_dataframe['dob'], errors='coerce')


fight_data['f_1'] = pd.to_numeric(fight_data['f_1'], errors='coerce')
fight_data['f_2'] = pd.to_numeric(fight_data['f_2'], errors='coerce')
fight_data['winner'] = pd.to_numeric(fight_data['winner'], errors='coerce')
fight_stat_data[['fighter_id', 'knockdowns', 'total_strikes_att', 
                 'total_strikes_succ', 'sig_strikes_att', 'sig_strikes_succ', 
                 'takedown_att', 'takedown_succ', 'submission_att', 'reversals']] = \
fight_stat_data[['fighter_id', 'knockdowns', 'total_strikes_att', 
                 'total_strikes_succ', 'sig_strikes_att', 'sig_strikes_succ', 
                 'takedown_att', 'takedown_succ', 'submission_att', 'reversals']].apply(pd.to_numeric, errors='coerce')


event_data_dtypes = event_data.dtypes
fighter_data_dtypes = fighter_data.dtypes
fight_data_dtypes = fight_data.dtypes
fight_stat_data_dtypes = fight_stat_data.dtypes
master_dataframe_dtypes = master_dataframe.dtypes

event_data_dtypes, fighter_data_dtypes, fight_data_dtypes, fight_stat_data_dtypes, master_dataframe_dtypes


(event_id                  int64
 event_name               object
 event_date       datetime64[ns]
 event_city               object
 event_state              object
 event_country            object
 event_url                object
 dtype: object,
 fighter_id                     int64
 fighter_f_name                object
 fighter_l_name                object
 fighter_nickname              object
 fighter_height_cm            float64
 fighter_weight_lbs           float64
 fighter_reach_cm             float64
 fighter_stance                object
 fighter_dob           datetime64[ns]
 fighter_w                      int64
 fighter_l                      int64
 fighter_d                      int64
 fighter_nc_dq                float64
 fighter_url                   object
 dtype: object,
 fight_id            int64
 event_id            int64
 referee            object
 f_1               float64
 f_2               float64
 winner            float64
 num_rounds         object
 title_fight    

### Inconsistent Data Entry
"orthadox" to "orthodox"

open stance = open_stance

women's weight class = w- 

In [57]:

fighter_data['fighter_stance'] = fighter_data['fighter_stance'].str.lower().replace(
    {'orthadox': 'orthodox', 'open stance': 'open_stance', 'sideways': 'sideways'})

fight_data['weight_class'] = fight_data['weight_class'].str.lower().replace(
    {"women's strawweight": "w-strawweight", 
     "women's bantamweight": "w-bantamweight",
     "women's flyweight": "w-flyweight",
     "women's featherweight": "w-featherweight",
     'catch weight': 'catch_weight', 
     'open weight': 'open_weight'})


master_dataframe['stance'] = master_dataframe['stance'].str.lower().replace(
    {'orthadox': 'orthodox', 'open stance': 'open_stance', 'sideways': 'sideways'})


fighter_stance_unique_cleaned = fighter_data['fighter_stance'].unique()
weight_class_unique_cleaned = fight_data['weight_class'].unique()
master_stance_unique_cleaned = master_dataframe['stance'].unique()

fighter_stance_unique_cleaned, weight_class_unique_cleaned, master_stance_unique_cleaned


(array(['not available', 'orthodox', 'switch', 'southpaw', 'open_stance',
        'sideways'], dtype=object),
 array(['lightweight', 'featherweight', 'w-strawweight', 'welterweight',
        'bantamweight', 'middleweight', 'heavyweight', 'w-bantamweight',
        'w-flyweight', 'flyweight', 'light heavyweight', 'catch_weight',
        'w-featherweight', 'open_weight', 'not available'], dtype=object),
 array([nan, 'orthodox', 'southpaw', 'open_stance', 'sideways', 'switch'],
       dtype=object))

### Filtering irrelevant colums 

Have not done yet, need to discuss what we do wanna keep and remove

But example:

In [58]:
event_data_cleaned = event_data.drop(columns=['event_url'])

fight_data_cleaned = fight_data.drop(columns=['fight_url', 'referee'])

fight_stat_data_cleaned = fight_stat_data.drop(columns=['fight_url'])

fighter_data_cleaned = fighter_data.drop(columns=['fighter_url', 'fighter_nickname'])

master_dataframe_cleaned = master_dataframe.drop(columns=['event_url', 'fight_url'])


event_data_cleaned_columns = event_data_cleaned.columns
fight_data_cleaned_columns = fight_data_cleaned.columns
fight_stat_data_cleaned_columns = fight_stat_data_cleaned.columns
fighter_data_cleaned_columns = fighter_data_cleaned.columns
master_dataframe_cleaned_columns = master_dataframe_cleaned.columns[:10]  # Display first 10 columns only

event_data_cleaned_columns, fight_data_cleaned_columns, fight_stat_data_cleaned_columns, fighter_data_cleaned_columns, master_dataframe_cleaned_columns


(Index(['event_id', 'event_name', 'event_date', 'event_city', 'event_state',
        'event_country'],
       dtype='object'),
 Index(['fight_id', 'event_id', 'f_1', 'f_2', 'winner', 'num_rounds',
        'title_fight', 'weight_class', 'gender', 'result', 'result_details',
        'finish_round', 'finish_time'],
       dtype='object'),
 Index(['fight_stat_id', 'fight_id', 'fighter_id', 'knockdowns',
        'total_strikes_att', 'total_strikes_succ', 'sig_strikes_att',
        'sig_strikes_succ', 'takedown_att', 'takedown_succ', 'submission_att',
        'reversals', 'ctrl_time'],
       dtype='object'),
 Index(['fighter_id', 'fighter_f_name', 'fighter_l_name', 'fighter_height_cm',
        'fighter_weight_lbs', 'fighter_reach_cm', 'fighter_stance',
        'fighter_dob', 'fighter_w', 'fighter_l', 'fighter_d', 'fighter_nc_dq'],
       dtype='object'),
 Index(['date', 'result', 'fighter', 'opponent', 'division', 'stance', 'dob',
        'method', 'total_comp_time', 'round'],
       dtype=

### Merge Datasets:

Again, need to discuss. 

example: 

Fight Data + Fight Stat Data: Merged on fight_id using inner join.

Merged with Fighter Data: Linked using f_1 as the fighter_id from Fighter Data (main fighter for the fight).

Used left join to ensure no missing fighters from the fight data.


In [59]:

fighter_data['fighter_f_name'] = fighter_data['fighter_f_name'].str.strip().str.lower()
fighter_data['fighter_l_name'] = fighter_data['fighter_l_name'].str.strip().str.lower()

fight_data_cleaned['fight_id'] = fight_data_cleaned['fight_id'].astype(int)
fight_stat_data_cleaned['fight_id'] = fight_stat_data_cleaned['fight_id'].astype(int)
fighter_data_cleaned['fighter_id'] = fighter_data_cleaned['fighter_id'].astype(int)

merged_fight_data = pd.merge(fight_data_cleaned, fight_stat_data_cleaned, on='fight_id', how='inner')

merged_fight_fighter_data = pd.merge(merged_fight_data, fighter_data_cleaned, left_on='f_1', right_on='fighter_id', how='left')


merged_shape = merged_fight_fighter_data.shape
merged_preview = merged_fight_fighter_data.head()

merged_shape, merged_preview


((14436, 37),
    fight_id  event_id     f_1     f_2  winner num_rounds title_fight  \
 0      7218       664  2976.0  2884.0  2884.0          5           F   
 1      7218       664  2976.0  2884.0  2884.0          5           F   
 2      7217       664  1662.0  2464.0  1662.0          3           F   
 3      7217       664  1662.0  2464.0  1662.0          3           F   
 4      7216       664   981.0   179.0   981.0          3           F   
 
     weight_class gender    result  ... fighter_l_name  fighter_height_cm  \
 0    lightweight      M    KO/TKO  ...         Fiziev             172.72   
 1    lightweight      M    KO/TKO  ...         Fiziev             172.72   
 2  featherweight      M  Decision  ...       Mitchell             177.80   
 3  featherweight      M  Decision  ...       Mitchell             177.80   
 4  w-strawweight      F    KO/TKO  ...      Rodriguez             167.64   
 
   fighter_weight_lbs  fighter_reach_cm  fighter_stance  fighter_dob  \
 0        

In [61]:
merged_fight_fighter_data.to_csv('merged_fight_fighter_data.csv', index=False)

### Saving Cleaned Data

In [62]:
event_data.to_csv('cleaned_event_data.csv', index=False)
fight_data.to_csv('cleaned_fight_data.csv', index=False)
fight_stat_data.to_csv('cleaned_fight_stat_data.csv', index=False)
fighter_data.to_csv('cleaned_fighter_data.csv', index=False)